src/NeoSharp.Core/Messaging/MessageHandlerProxy.cs

```diff
@@ -1,45 +1,138 @@
             _logger = logger;
-            _messageHandlerInvokers = messageHandlerTypes
-                .Select(CreateMessageHandlerInvoker)
-                .ToDictionary(x => x.MessageType, x => x.MessageHandlerInvoker);
+            _messageHandlerTypes = messageHandlerTypes.ToArray();
         }
 
+        /// <summary>
+        /// Handle message
+        /// </summary>
+        /// <param name="message">Message</param>
+        /// <param name="sender">Sender</param>
+        /// <returns>Task</returns>
         public Task Handle(Message message, IPeer sender)
         {
             if (message == null) throw new ArgumentNullException(nameof(message));
 
-            var messageType = message.GetType();
-            var messageHandler = ResolveMessageHandler(messageType);
-            var startedAt = LogMessageHandlingStart(messageHandler);
-            var messageHandlerInvoker = GetMessageHandlerInvoker(messageType);
-            var handleMessageTask = (Task)messageHandlerInvoker.DynamicInvoke(messageHandler, message, sender);
+            if (_reflectionCache == null)
+            {
+                // Create cache
+
+                var messageHandlerInvokers = _messageHandlerTypes.Select(CreateMessageHandlerInvoker)
+                    .ToDictionary(x => x.MessageType, x => x.MessageHandlerInvoker);
+
+                byte max = 0;
+                var cache = ReflectionCache<MessageCommand>.CreateFromEnum<MessageCommand>();
+                var r = new Cache[byte.MaxValue];
+
+                foreach (MessageCommand v in Enum.GetValues(typeof(MessageCommand)))
+                {
+                    if (!cache.TryGetValue(v, out Type centry)) continue;
+
+                    byte val = (byte)v;
+
+                    if (!messageHandlerInvokers.TryGetValue(centry, out var messageHandlerInvoker))
+                    {
+                        throw new InvalidOperationException(
+                            $"The message of \"{centry}\" type has no registered handlers.");
+                    }
+
+                    r[val] = new Cache(v, _container.Resolve(typeof(IMessageHandler<>).MakeGenericType(centry)), messageHandlerInvoker);
+                    max = Math.Max(max, val);
+                }
+
+                Array.Resize(ref r, max + 1);
+                _reflectionCache = r;
+            }
+
+            // Extract handler
+
+            var entry = _reflectionCache[(byte)message.Command];
+
+            if (entry == null)
+                throw new InvalidOperationException($"The message of \"{ message.GetType()}\" type has no registered handlers.");
+
+            // Log start
+
+            var startedAt = DateTime.Now;
+            _logger.LogDebug($"The message handler \"{entry.MessageHandlerName}\" started message handling at {startedAt:yyyy-MM-dd HH:mm:ss}.");
+
+            // Execute
+
+            var handleMessageTask = (Task)entry.HandlerInvoker.DynamicInvoke(entry.MessageHandler, message, sender);
 
```

Comment:
I did not get why we need to this complexity. Could you elaborate what you are trying to achieve?

Reply:
Now we don't need to get 3 objects from different caches, now all is cached in only one object

Reply:
The readability cost is to high for this change. Meanwhile, it creates cache on every ```Handle``` call but ```Handle``` should operate already on warmed up cache. Do you mind restore old version?

Reply:
Really i think that is easier to read now, before the change a lot of calls was made per received message. Now all the stuff related to one command are on the same place. What is the difficult?

Reply:
All these calls are cheap and will be inlined by the compiler. So the goal was to make it readable. I remember you mentioned on the channel it was unreadable when the method contained ~20 lines. Now it is 63 lines. I am still trying to wrap my head around what specifically does this code
```max = Math.Max(max, val);```
or this
```Array.Resize(ref r, max + 1);
_reflectionCache = r;```
How is it related to message handling? I wrote the previous version of the code and it is hard for me to read this version. I can imagine what if @aboimpinto decides to write tests for this code how many challenges he will probably face.

Reply:
I don't like methods that only will be called one time, in critical part of the code like this

Reply:
I will repeat this again. These method calls do not exist in compiled code: they are inlined. We do not do compiler work we write code for humans. Any method that contains more than 10 lines has to be broken down.

Reply:
I agree with @osmirnov let's keep the method line size down and the code human readable... as stated it all compiles to the same anyway so no performance hit.

Reply:
what code exactly you said that does not exist in compiled code? The problem here is that we need a rules, because the previous code is less readable for me, an this is less readable for you, and both are humans xD, so we need to define the rules together for prevent this isues

Reply:
That's why we need to involve more people into our discussions :smile:

Reply:
A few points we want to keep in mind here:

1. The focus of neo-sharp is to create a modular, maintainable codebase by multiple collaborators, the focus initially is not on performance - that will come in future iterations and optimizations, only as necessary.  I can assure you that the performance issues we currently have in NEO will not be solved by minor code differences at this level - it will be sweeping pattern changes or innovations.  

2.  Human readability and maintainability is key - if we want to get lots of contributors on the project, the code has to be readable.  

To these points, I am 100% in agreement with @osmirnov and @AshRolls  - bloated objects, functions, and complex / hard to read logic, even if "more performance" is actually a negative to the project overall  if it makes code less readable and maintainable.

Reply:
Could you explain me what is the unreadable part compared with the previous version?

Reply:
Because i only put this together in the same class for prevent this calls for each message

https://github.com/CityOfZion/neo-sharp/blob/0bb524d5854d3c8ca52389883ebe7b34efc208b7/src/NeoSharp.Core/Messaging/MessageHandlerProxy.cs#L63

https://github.com/CityOfZion/neo-sharp/blob/0bb524d5854d3c8ca52389883ebe7b34efc208b7/src/NeoSharp.Core/Messaging/MessageHandlerProxy.cs#L48-L58

https://github.com/CityOfZion/neo-sharp/blob/0bb524d5854d3c8ca52389883ebe7b34efc208b7/src/NeoSharp.Core/Messaging/MessageHandlerProxy.cs#L32-L36

New version

https://github.com/shargon/neo-sharp/blob/fcb5ec5cf7a172101dcef16e81fe2943a58cfbdc/src/NeoSharp.Core/Messaging/MessageHandlerProxy.cs#L77

and the generation is the same but in a loop

https://github.com/shargon/neo-sharp/blob/fcb5ec5cf7a172101dcef16e81fe2943a58cfbdc/src/NeoSharp.Core/Messaging/MessageHandlerProxy.cs#L135-L166

But tell me what part is ugly, and i try to change it.

Reply:
@shargon I see you took most of all my suggestions but the part with cache is still very confusing. Moreover, you've altered behavior.
In the old version, the handler was intentionally created per incoming message. It allowed supporting them as stateless, avoiding any memory upholding and activating IDisposal mechanism.
In the new version, all message handlers are singleton and stateful that can lead to all mentioned problems.
Anyway.
You said we cannot decide on readability and I invited more team members to take a look. @AshRolls and @gubanotorious (thanks guys!) reviewed that piece of the code and agreed that became less readable. Could we revert it now?

Reply:
Please specify what part do you want to revert